packages/accounts-controller/src/AccountsController.ts

```diff
@@ -1169,6 +1135,90 @@ export class AccountsController extends BaseController<
     return internalAccount ? internalAccount.metadata[metadataKey] : undefined;
   }
 
+  /**
+   * Subscribes to message events.
+   * @private
+   */
+  #subscribeToMessageEvents() {
+    this.messagingSystem.subscribe(
+      'SnapController:stateChange',
+      (snapStateState) => this.#handleOnSnapStateChange(snapStateState),
+    );
+
+    this.messagingSystem.subscribe(
+      'KeyringController:stateChange',
+      (keyringState) => this.#handleOnKeyringStateChange(keyringState),
+    );
+
+    this.messagingSystem.subscribe(
+      'SnapKeyring:accountAssetListUpdated',
+      (snapAccountEvent) =>
+        this.#handleOnSnapKeyringAccountEvent(
+          'AccountsController:accountAssetListUpdated',
+          snapAccountEvent,
+        ),
+    );
+
+    this.messagingSystem.subscribe(
+      'SnapKeyring:accountBalancesUpdated',
+      (snapAccountEvent) =>
+        this.#handleOnSnapKeyringAccountEvent(
+          'AccountsController:accountBalancesUpdated',
+          snapAccountEvent,
+        ),
+    );
+
+    this.messagingSystem.subscribe(
+      'SnapKeyring:accountTransactionsUpdated',
+      (snapAccountEvent) =>
+        this.#handleOnSnapKeyringAccountEvent(
+          'AccountsController:accountTransactionsUpdated',
+          snapAccountEvent,
+        ),
+    );
+
+    // Handle account change when multichain network is changed
+    this.messagingSystem.subscribe(
+      'MultichainNetworkController:setActiveNetwork',
```

Comment:
Should this be '
`MultichainNetworkController:stateChange`?

I think here we subscribe to the action which can fail, but if we listen to the state change we are sure that the account needs to switch! Wdyt?